core/objects.go

```diff
@@ -103,8 +110,9 @@ const DNSPrefix = "_acme-challenge"
 // addresses, etc.), but currently we only support
 // domain names.
 type AcmeIdentifier struct {
-	Type  IdentifierType `json:"type"`  // The type of identifier being encoded
-	Value string         `json:"value"` // The identifier itself
+	Type     IdentifierType `json:"type"`  // The type of identifier being encoded
+	Value    string         `json:"value"` // The identifier itself
+	Wildcard bool           `json:"-"`     // Is the value of the identifier the base domain of a wildcard?
```

Comment:
This fields gets set when building an order, but it doesn't get serialized to disk because, as we discussed, we rely on the index over the JSON-serialized `identifier` field, and adding a new field would break this.

The lack of serialization creates a problem when identifiers are read back. How do we set the Wildcard field when we read an authorization out of pendingAuthorizations or authz? We can say "if an authorization object has exactly one challenge, and that challenge is of type "dns-01-wildcard", we can set the Wildcard bit on its identifier. However, this suggests what might be an easier, clearer path: Make wildcard-ness a property of authorizations, and rather than represent it as a boolean field, represent it with a `.Wildcard()` method on authz that calculates wildcard-ness based on the set of challenges. What do you think?

Reply:
This sounds like a nice improvement. I'll give it a shot. 

I think it will end up meaning we have to pass an authorization to the PA to get challenges - right now it only accepts an identifier. If the authz is marked as a wildcard intended to authorize "*.example.com" but the identifier only has "example.com" with no wildcard marking when we ask for challenges we won't know to only construct DNS-01-Wildcard.

Reply:
After more thought I'm not sure this works very well. You still need a way to know that a translated (e.g. base domain, no `*`) identifier was for a wildcard. 

E.g. consider in `new-order` when we have to check whether we have any authorizations that can be reused for the names/identifiers in the order. If the order was for `"*.example.com"` then we have a DNS type identifier with the value `"example.com"`. We can look up the existing authz's and call `Wildcard()` on each, but we don't know whether to care if the answer is true or false unless we know that the `"example.com"` identifier value is for a wildcard order name `"*.example.com"`. 

Without carrying forward the Wildcard information on the identifier it becomes tricky to measure "Here's a set of Authorizations for some Identifier Values, some of the authorizations are Wildcard()" against "Here's a set of Order names with wildcard characters, Here's the set of Identifiers Values that need to have authorizations". You have to match the identifier value `example.com` against an Authz for `example.com` that has Wildcard() but only when you are dealing with the order name `*.example.com`.

I think perhaps the best follow-up is to switch to the scheme @rolandshoemaker proposed (I think?) in a sprint call where we allow the identifier to take the literal value `"*.example.com"` and then also adopt your `authz.Wildcard()` method. We would have to translate the identifier back to `example.com` for presentation to the user to obey ACME identifier semantics. I'm open to this but its more or less a rewrite of this PR.

Reply:
> I think perhaps the best follow-up is to switch to the scheme @rolandshoemaker proposed (I think?) in a sprint call where we allow the identifier to take the literal value "*.example.com" and then also adopt your authz.Wildcard() method. We would have to translate the identifier back to example.com for presentation to the user to obey ACME identifier semantics. I'm open to this but its more or less a rewrite of this PR.

I think this makes sense and has an appealing simplicity. One of the implications is that we won't (in general) use wildcard authorizations for non-wildcard issuance, and vice versa. That is, if you `new-order(example.com)`, and happen to fulfill a DNS challenge for it, then later `new-order(*.example.com)`, we won't use the cached authorization but will create a new one.

In general, I think this is fine. Just to mention: `new-order(*.example.com, example.com)` is going to be very common. Ideally we'd like to create just one authorization for that. We could do that by jumping through some hoops in new-order to create just one (the wildcard one), and at issuance time, consider that good enough to issue for both. However, this would make checking CAA really complicated, since we currently check it at validation time. So I would vote to punt on this issue for now (and maybe double-check that two DNS challenge responses can be live on a given domain at once).

Reply:
I was going to push a commit that adds the `core.Authorization.Wildcard()` method but also leaves the `core.AcmeIdentifier.Wildcard` field to try and clean up the RA code.

This turns out to be a bit annoying too because presently the RA deals with `core.proto.Authorization` instances in some places and I don't think there's an easy way to add the method to that type and it doesn't seem to justify juggling back to the core type.

Thoughts?

Reply:
Now that we are planning to stick the `*` in the identifier, I think we can go back to treating wildcard-ness as a property of identifiers. To get around the method problem, we can define an `IsIdentifierWildcard(core.proto.AcmeIdentifier)` method, and a corresponding one that works on `core.AcmeIdentifier`, right?

Reply:
> So I would vote to punt on this issue for now (and maybe double-check that two DNS challenge responses can be live on a given domain at once).

Sounds good. I will verify that this is the case and we can punt on a cleaner design and trying to avoid duplicating an authz for `new-order(*.example.com, example.com)`

Reply:
>  I will verify that this is the case and we can punt on a cleaner design and trying to avoid duplicating an authz for new-order(*.example.com, example.com)

The VA uses the `bdns` `LookupTXT` function to fetch TXT records for a name. [That function returns multiple TXTs if present](https://github.com/letsencrypt/boulder/blob/72330bbedd201c9c23365840fe28ec76e7bba3ed/bdns/dns.go#L338:L366). The VA **does** [check each TXT record against the expected](https://github.com/letsencrypt/boulder/blob/72330bbedd201c9c23365840fe28ec76e7bba3ed/va/va.go#L756:L764).

It shouldn't be a problem to have two DNS-01 TXT records for `example.com`, one for `*.example.com` and one for `example.com`. The VA will handle it as-expected.